### DISCUSSION TITLE
Applying syscall parameters and semantics to an ARM SWI instruction

### DISCUSSION THREAD

**Discussion by philpem** [NationalSecurityAgency/ghidra#7274] [Q&A]
I'm working on a piece of code for Acorn RISC OS. This is a 1990s era (though updated to present day for the Raspberry Pi) operating system for the ARM platform.

OS syscalls are done with a SWI instruction, e.g.:

```
        0000c39c 66 03 00 eb     bl         ClearR1ToR7                                      undefined ClearR1ToR7()
        0000c3a0 01 30 a0 e3     mov        r3,#0x1
        0000c3a4 00 20 a0 e1     mov        r2,r0
        0000c3a8 14 10 9b e5     ldr        r1,[r11,#0x14]
        0000c3ac 55 00 a0 e3     mov        r0,#0x55
        0000c3b0 3a 0c 80 e3     orr        r0,r0,#0x3a00
        0000c3b4 6e 08 80 e3     orr        r0,r0,#0x6e0000
        0000c3b8 81 04 80 e3     orr        r0,r0,#0x81000000
        0000c3bc 5b 00 00 ef     swi        OS_CRC
        0000c3c0 38 03 8c e5     str        r0,[r12,#0x338]
        0000c3c4 10 10 9b e5     ldr        r1,[r11,#0x10]
        0000c3c8 38 03 9c e5     ldr        r0,[r12,#0x338]
```

Annoyingly this appears in the decompiler as:

```
software_interrupt(OS_CRC);
```

Is there a way to tell Ghidra about the calling conventions for the various syscalls, similar to how they can be specified with functions?

My goal is to write a script which decodes the SWI numbers (by creating enumerations) and applies the calling convention hints for the syscalls it knows about.

**Comment by astrelsky** [comment]
This should help
https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/ghidra_scripts/ResolveX86orX64LinuxSyscallsScript.java

**Comment by hippietrail** [comment]
I'm interested in this kind of issue generally for all kinds of BIOS/OS calls using instructions such as `int`, `trap`, `svc`, `swi`, `ecall`, etc. on various CPU architectures.

**Reply by ABratovic** [reply]
@hippietrail 
Hi Andrew
have a look at https://github.com/Gravelbones/GhidraDosToolbox 
I found it in the comments for https://github.com/NationalSecurityAgency/ghidra/issues/2266
